Add first stage of ocean/ice shelf coupling

matthewhoffman opened the pull request:
This merge introduces coupler changes to allow coupling between the ocean model and ice shelves within the GLC model.

As part of model coupling necessary for Antarctic simulations, ice shelf melt fluxes due to (primarily) ocean temperatures must be calculated.  These melt fluxes correspond to oceanic heat and moisture fluxes at the upper ocean surface, and ice sheet heat and ice column thickness changes.  Coupling is characterized by 1) calculations carried out by ‘subroutine compute_melt_fluxes’ on the GLC grid, but 2) on the OCN coupling timestep (<GLC coupling timestep).  This is why these fluxes cannot be calculated in either component and are done here in the coupler itself.  These factors result in 1) frequent OCN2GLC remapping that is not associated with a GLC run step and 2) accumulation of GLC-bound ‘subroutine compute_melt_fluxes’ output fields, but no equivalent accumulation of OCN-bound outputs.

Inputs to the flux calculations (all mapped to GLC grid):
From OCN:
*   oceanTemperature(:) -> o2x_gx%rAttr(index_o2x_So_blt,:)
*  oceanSalinity(:) -> o2x_gx(1)%rAttr(index_o2x_So_bls,:)
*   oceanHeatTransferVelocity(:) -> o2x_gx(1)%rAttr(index_o2x_So_htv,:) 
*   oceanSaltTransferVelocity(:) -> o2x_gx(1)%rAttr(index_o2x_So_hsv,:)

From GLC:
*   interfacePressure(:) -> o2x_gx(1)%rAttr(index_o2x_So_phieff,:)
*   iceTemperature(:) -> g2x_gx%rAttr(index_g2x_Sg_tbot,:)
*   iceTemperatureDistance(:) -> g2x_gx%rAttr(index_g2x_Sg_dztbot,:)
*   iceFloatingMask(:) -> g2x_gx%rAttr(index_g2x_Sg_icemask_floating,:)

Outputs from flux calculations (all calculated natively on GLC grid):
To OCN:
*   g2x_gx%rAttr(index_g2x_Sg_blis,:) -> outInterfaceSalinity(:)
*   g2x_gx%rAttr(index_g2x_Sg_blit,:) -> outInterfaceTemperature(:) 
*   g2x_gx%rAttr(index_g2x_Fogx_qiceho,:) -> outOceanHeatFlux(:) 
*   g2x_gx%rAttr(index_g2x_Fogx_qicelo,:) -> outFreshwaterFlux(:)

TO GLC:
*   x2g_gx%rAttr(index_x2g_Fogx_qicehi,:) -> outIceHeatFlux(:)
*   x2g_gx%rAttr(index_x2g_Fogx_qiceli,:) -> -1*outFreshwaterFlux(:)

These changes also include necessary driver and build script changes and a new compset "GMPAS-MALI-DIB-IAF-ISMF” that includes active MPAS-Ocean and MPAS-Albany-Land-Ice (as well as active MPAS-Sea-Ice).  In this compset, MPAS-Albany-Land-Ice is configured in a ’static’ mode where ice sheet geometry and temperature do not evolve.

Changes are BFB for existing compsets.

matthewhoffman commented on components/mpas-albany-landice/driver/glc_cpl_indices.F:
```diff
@@ -14,25 +14,34 @@ module glc_cpl_indices
-  integer, public :: index_x2g_Flgl_qice = 0 !Jer: removed multiple fields, now that these are confined to coupler
-  integer, public :: index_x2g_Fogo_qicel = 0 
-  integer, public :: index_x2g_Sl_tsrf   = 0 !Jer: removed multiple fields, now that these are confined to coupler
-  integer, public :: index_x2g_Fogo_qiceh = 0 !Jer:need to rename this everywhere...
```
@JeremyFyke , what was x2g_Fogo_qiceh and why has it been removed/renamed?

JeremyFyke replied:
By the name, this was a heat flux field from coupler to glacier (originally from the ocean, I think).  My nominal guess here is that I renamed this to something else, or realized it was not needed.

matthewhoffman commented on cime/src/drivers/mct/cime_config/config_component_e3sm.xml:
```diff
@@ -781,6 +781,40 @@
     <desc>glc2ocn state mapping file decomp type</desc>
   </entry>
 
+  <entry id="GLC2ICE_FMAPNAME">
```
@JeremyFyke , @rljacob, @jonbob ,  Jeremy's original changes had these added to config_component.xml.  What is the difference between config_component.xml and config_component_e3sm.xml and which file should these changes be added to?

rljacob replied:
Its possible that config_component_e3sm.xml didn't exist when Jeremy did this work.
The difference is the *e3sm* one is for driver config items that are either unique to E3SM or require E3SM-specific defaults.

JeremyFyke replied:
I don't recall a config_component_e3sm.xml so I think Rob is correct here.

matthewhoffman commented on cime/src/drivers/mct/main/cime_comp_mod.F90:
```diff
@@ -410,6 +411,8 @@ module cime_comp_mod
   logical  :: glc_c2_lnd             ! .true.  => glc to lnd coupling on
   logical  :: glc_c2_ocn             ! .true.  => glc to ocn coupling on
   logical  :: glc_c2_ice             ! .true.  => glc to ice coupling on
+  logical  :: glcshelf_c2_ocn        ! .true.  => glc ice shelf to ocn coupling on
```
@JeremyFyke , what is the intended difference between glc_c2_ocn/ice and glcshelf_c2_ocn/ice?  I thought there was no existing GLC to OCN/ICE coupling.

JeremyFyke replied:
glc_c2_ocn and glc_c2_ice originally (and I think still) relate to iceberg calving routing.  So, I made a new coupler flag to differentiate and separate these couplings (with unique mapping files, etc.) from all the subshelf melting couplings (glcshelf_c2_ocn).

matthewhoffman commented on cime/src/drivers/mct/main/prep_ocn_mod.F90:
```diff
@@ -1066,6 +1110,8 @@ subroutine prep_ocn_merge( flux_epbalfact, a2x_o, i2x_o, r2x_o, w2x_o, g2x_o, xa
                x2o_o%rAttr(index_x2o_Faxa_snow_HDO ,n)
        end if
     end do
+    !Experimental: call subshelf boundary flux code here.  Note this is called on atmospheric timestep.
```
@JeremyFyke , can you elaborate on this comment?  What is experimental about this?  Why is it on the atmospheric timestep?  Would it ideally be called on the ocean coupling timestep?

JeremyFyke replied:
I *think* perhaps it could be called on the ocean coupling timestep, equivalently.  I believe 'experimental' refers to that of all this development, the impact of coupler timesteps was/is the least clear to me.  I think, for example, that my implementation may be doing redundant accumulation/averaging steps in the coupler.  Also, I didn't test different coupler timestep schemes (e.g. as explored by @proteanplanet)  @rljacob - I think it would be really worthwhile for you to take a look at the implementation here to see if you see any errors.

matthewhoffman commented on cime/src/share/util/shr_const_mod.F90:
```diff
@@ -54,6 +54,12 @@ MODULE shr_const_mod
+   real(R8),parameter :: SHR_CONST_CONDICE = 2.1_R8          ! thermal conductivity of ice ~ W/m/K
+   real(R8),parameter :: SHR_CONST_KAPPA_LAND_ICE = &        ! Diffusivity of heat in land ice ~
+                         SHR_CONST_CONDICE / (SHR_CONST_RHOICE*SHR_CONST_CPICE)
+   real(R8),parameter :: SHR_CONST_DTF_DP = -7.53e-8_R8      !Jer TODO: add on description
```
@JeremyFyke , what are these new constants?

matthewhoffman commented on cime/src/drivers/mct/main/prep_glc_mod.F90:
```diff
@@ -128,14 +170,19 @@ subroutine prep_glc_init(infodata, lnd_c2_glc)
 
     smb_renormalize = prep_glc_do_renormalize_smb(infodata)
 
+    g2x_gx => component_get_c2x_cx(glc(1)) !Jer: is this needed here?
```
@JeremyFyke , can you elaborate on what the potential issue is here?

JeremyFyke replied:
I am not sure, sorry.  Perhaps nothing from this structure is utilized in following code?

matthewhoffman replied:
It appears to be unused, so I removed it.

matthewhoffman commented on cime/src/drivers/mct/main/prep_glc_mod.F90:
```diff
@@ -342,6 +458,21 @@ subroutine prep_glc_accum_avg(timer)
+	  call mct_avect_avg(x2gacc_gx(egi), x2gacc_gx_cnt)
+       end if
+
+       ! ***NOTE***THE FOLLOWING ACTUALLY MODIFIES x2g_gx
```
@JeremyFyke , this comment doesn't sound good.  Do we need to investigate a different implementation here?

JeremyFyke replied:
Can you grep for this comment string elsewhere?  I *suspect* it is not unique to this occurrence.  If not, I'll investigate further.   While it would be good to do a 'proper' implementation, I would ask @rljacob to check this.  It may not be STRICTLY necessary, as I recall (otherwise I would have done it, in theory)

rljacob replied:
What is the glacier model coupling frequency?  Nearly all the other components, including ocean, are coupling every hour now.

matthewhoffman replied:
@rljacob, right now we have GLC coupling once per day.  It could be longer (e.g. a week or two), but a day is easier to deal with for now.  The disadvantage to making it shorter is that the ice sheet time step is order weeks, so a really small coupling interval would require running the ice sheet model much, much more frequency than necessary, at associated expense.

rljacob commented on cime/src/drivers/mct/main/prep_glc_mod.F90:
```diff
@@ -607,6 +765,149 @@ end subroutine prep_glc_map_one_state_field_lnd2glc
+    g2x_gx => component_get_c2x_cx(glc(1))
+    x2g_gx => component_get_x2c_cx(glc(1))
+
+    !TO DO: CONSIDER HOW ENSEMBLES ARE GOING TO BE IMPLEMENTED (i.e. change hard-coded '1' index to e*i indices
```
Yes.  See the similar code in in prep_ocn_mod.F90

JeremyFyke replied:
As Rob mentioned.  I put this off in interests of getting basic scheme operating.

rljacob replied:
You could also look at the latest prep_glc_mod.F90 on the cime master where the ensemble pieces are added.

matthewhoffman commented on cime/src/drivers/mct/main/prep_glc_mod.F90:
```diff
@@ -196,6 +244,69 @@ subroutine prep_glc_init(infodata, lnd_c2_glc)
+       allocate(outFreshwaterFlux(lsize_g))
+       allocate(outOceanHeatFlux(lsize_g))
+       allocate(outIceHeatFlux(lsize_g))
+       ! TODO: Can we allocate these only while used or are we worried about performance hit?
```
@rljacob , is it preferable to 1) allocate these once on init and keep the memory allocated for the entire simulation but only have one allocation or 2) allocate/deallocate these at each coupling interval which would have smaller memory footprint but require many allocate operations?

matthewhoffman replied:
@rljacob , @jgfouca , do either of you have a preference for the implementation of allocate/deallocate here?

rljacob replied:
This routine is only called once per coupling step so it probably won't matter for timing.  If those arrays are huge, you should probably allocate once.

matthewhoffman replied:
@rljacob , I don't think they will ever be huge.  Their length is the number of GLC grid cells that overlap with the OCN grid.  But if it doesn't matter much either way, I'll leave the current implementation for simplicity.

matthewhoffman commented:
As of 53259cdad96864e5541a18013b02b3afea0534ca, the subshelf melt calculation yields the correct behavior (with some exceptions).  I set up E3SM/MPASO to perform the melt calculations both inside MPASO and inside the E3SM coupler.  MPASO only sees the melt calculation that is calculated in MPASO, so the E3SM melt calculation is effectively diagnostic using the same state fields as the MPASO melt calculation.  

This figure show that the melt flux calculated in E3SM matches the MPASO version.  Results are from 01-01-04 of a simulation starting on 01-01-01.
* Bottom left is the MPAS version on the OCN grid
* Bottom right is E3SM version on the OCN grid
* Upper left is E3SM version on the GLC grid
![subshelf-melt-comparison](https://user-images.githubusercontent.com/4182034/53508172-78474780-3a76-11e9-9281-ffdd5ae8356e.png)

There still are some minor issues to resolve:
* The melt flux on the GLC grid shows some unrealistically high values along edges of the ice shelves.  Presumably this is due to some sort of masking/partial-overlap issue.
* The melt calculation is invalid for the first coupling interval.  The simulation starts on 01-01-01 with a OCN-GLC coupling interval of 1 day and the first time that the melt calculation is correct in the coupler history files is 01-01-03.  There is presumably a lag in the coupling, but we'll have to figure out how to handle that best.

---
_For reference, I set this up by setting:_
```
config_land_ice_flux_mode = 'standalone'
```
_with these MPASO source modifications (thanks @xylar):_
```
diff --git a/src/core_ocean/driver/mpas_ocn_core_interface.F b/src/core_ocean/driver/mpas_ocn_core_interface.F
index f878167..fb914dd 100644
--- a/src/core_ocean/driver/mpas_ocn_core_interface.F
+++ b/src/core_ocean/driver/mpas_ocn_core_interface.F
@@ -254,6 +254,7 @@ module ocn_core_interface
       else if ( trim(config_land_ice_flux_mode) == 'standalone' ) then
          landIcePressurePKGActive = .true.
          landIceFluxesPKGActive = .true.
+         landIceCouplingPKGActive = .true.
       else if ( trim(config_land_ice_flux_mode) == 'coupled' ) then
          landIcePressurePKGActive = .true.
          landIceFluxesPKGActive = .true.
diff --git a/src/core_ocean/shared/mpas_ocn_time_average_coupled.F b/src/core_ocean/shared/mpas_ocn_time_average_coupled.F
index 6e89701..7f5247b 100644
--- a/src/core_ocean/shared/mpas_ocn_time_average_coupled.F
+++ b/src/core_ocean/shared/mpas_ocn_time_average_coupled.F
@@ -99,7 +99,7 @@ module ocn_time_average_coupled
         !$omp end do

         call mpas_pool_get_config(ocnConfigs, 'config_land_ice_flux_mode', config_land_ice_flux_mode)
-        if(trim(config_land_ice_flux_mode) == 'coupled') then
+!        if(trim(config_land_ice_flux_mode) == 'coupled') then
            call mpas_pool_get_array(forcingPool, 'avgLandIceBoundaryLayerTracers', avgLandIceBoundaryLayerTracers)
            call mpas_pool_get_array(forcingPool, 'avgLandIceTracerTransferVelocities', avgLandIceTracerTransferVelocities)
            call mpas_pool_get_array(forcingPool, 'avgEffectiveDensityInLandIce', avgEffectiveDensityInLandIce)
@@ -111,7 +111,7 @@ module ocn_time_average_coupled
               avgEffectiveDensityInLandIce(iCell) = 0.0_RKIND
            end do
            !$omp end do
-        end if
+!        end if

         !  set up BGC coupling fields if necessary
         call mpas_pool_get_config(ocnConfigs, 'config_use_ecosysTracers', config_use_ecosysTracers)
@@ -281,7 +281,7 @@ module ocn_time_average_coupled
         !$omp end do

         call mpas_pool_get_config(ocnConfigs, 'config_land_ice_flux_mode', config_land_ice_flux_mode)
-        if(trim(config_land_ice_flux_mode) == 'coupled') then
+!        if(trim(config_land_ice_flux_mode) == 'coupled') then
            call mpas_pool_get_array(diagnosticsPool, 'landIceBoundaryLayerTracers', landIceBoundaryLayerTracers)
            call mpas_pool_get_array(diagnosticsPool, 'landIceTracerTransferVelocities', landIceTracerTransferVelocities)
            call mpas_pool_get_array(statePool, 'effectiveDensityInLandIce', effectiveDensityInLandIce, timeLevel)
@@ -300,7 +300,7 @@ module ocn_time_average_coupled
                  + effectiveDensityInLandIce(iCell) ) / ( nAccumulatedCoupled + 1)
            end do
            !$omp end do
-        end if
+!        end if

         !  accumulate BGC coupling fields if necessary
         call mpas_pool_get_config(ocnConfigs, 'config_use_ecosysTracers', config_use_ecosysTracers) 
```
_I also added this stream to the `streams.ocean` file to see the relevant input and output fields for the melt fluxes calculated in MPASO to compare against:_
```
<stream name="forcingGLC"
        type="output"
        io_type="pnetcdf"
        filename_template="mpaso.hist.forcing_GLC_variables.$Y-$M-$D_$S.nc"
        filename_interval="00-00-01_00:00:00"
output_interval="00-00-01_00:00:00"
        clobber_mode="truncate">
    <var name="landIceSurfaceTemperature"/>
    <var name="landIceMask"/>
    <var name="effectiveDensityInLandIce"/>
    <var name="landIceDraft"/>
    <var name="landIcePressure"/>
    <var name="landIceFreshwaterFlux"/>
    <var name="landIceHeatFlux"/>
    <var name="landIceFraction"/>
 <var_array name="landIceInterfaceTracers"/>
<var_array name="landIceBoundaryLayerTracers"/>
<var_array name="landIceTracerTransferVelocities"/>
<var_array name="avgLandIceBoundaryLayerTracers"/>
<var_array name="avgLandIceTracerTransferVelocities"/>
<var_array name="avgEffectiveDensityInLandIce"/>
</stream>
```
---

matthewhoffman commented:
Update on the two issues noted above in my long comment from Feb. 27 (issue with values around edges and issue with values on first time step).

Both issues appear to be spin-up related effects and not problems with the implementation.  

* values around edges: These are places where the MPAS-O initial condition did not include ice shelves, but the coupled model calculation does due to a slightly different extent of the ice shelves in MALI.  Thus in MPAS-O the water is unphysically warm for an interface with ice and there are extreme melt rates calculated.  Tests are ongoing to see if these places will calm down on their own in a coupled run, or if more careful spin-up of MPAS-O will be necessary to eliminate this.
* values at initial coupling interval:  This is a similar problem.  If the OCN-GLC coupling interval is set to 30 minutes, then the E3SM calculation matches that of standalone MPAS-O.  However, when it is set to the default of 1 day, then we only see output for the accumulated value at 1 day, which includes extreme values from the first few time steps.

jonbob commented:
@matthewhoffman - I'm looking at the code, and the computation of melt fluxes is only occurring for cells with floating land ice. When we looked, the places with unphysical values also corresponded to cells without floating land ice? Am I remembering that correctly?

matthewhoffman commented:
@jonbob , what we saw was the place with large melt values were places that are floating ice in the coupled model (as dictated by the extent of ice shelves on the GLC grid), but did not contain ice shelves in the MPASO initial condition file.

matthewhoffman commented:
@jgfouca, I'm adding you as a reviewer due to the substantial changes to CIME (10 files, ~1200 lines).   Note that it's not clear how these changes will impact NCAR's use of CIME.  

The current branch is mostly complete, but there will still be some cleanup.

matthewhoffman commented:
I've pushed two new clean-up commits addressing issues discussed above.  The branch is ready for code review.

@rljacob , note that an overview of the flux calculation is here:
https://docs.google.com/drawings/d/1O_SBNWl6zuVINOLAk_L9ioHg4qFLeyA9k10FQO2HVeA/edit

And an overview of the code changes is here:
https://docs.google.com/drawings/d/1IbY-HnX24zV4jm5aaKBhORPg2cNF0v2hl0K5Mx4Bjgc/edit

jonbob commented:
I confirmed that a test merge of this PR does not impact results from configurations without active glc. Tests of SMS.ne30_oECv3_ICG.A_WCYCL1850S_CMIP6.anvil_intel were bfb, before and after the merge.

billsacks commented:
@rljacob thanks for pointing me to this; I'll try to at least give it a quick look to evaluate its impact on CESM.

billsacks commented:
An initial comment / request before I look at this more carefully: It looks like you (probably accidentally in some merge) removed the call to cime_run_glc_setup_send, instead inlining (and changing) the code from that routine. Can you please update this PR to call cime_run_glc_setup_send, merging your changes with that routine where necessary? Then I'll be able to review your changes to cime_comp_mod more easily.

billsacks commented on cime/src/drivers/mct/main/prep_glc_mod.F90:
```diff
@@ -607,6 +777,156 @@ end subroutine prep_glc_map_one_state_field_lnd2glc
+
+    do egi = 1,num_inst_glc
+
+    o2x_ox => component_get_c2x_cx(ocn(egi))
```
Please clean up the indentation of this routine.

matthewhoffman replied:
Fixed in 733a46925b5580e5888bdfa417add9916e7cf2df

billsacks commented on cime/src/drivers/mct/main/prep_glc_mod.F90:
```diff
@@ -607,6 +777,156 @@ end subroutine prep_glc_map_one_state_field_lnd2glc
+    !---------------------------------------------------------------
+
+    if (.not.(glc_present)) return
+    !write(logunit,*) 'Starting prep_glc_calculate_subshelf_boundary_fluxes'
```
Please remove commented-out lines of code like this.

matthewhoffman replied:
Fixed in 733a46925b5580e5888bdfa417add9916e7cf2df

matthewhoffman commented:
@billsacks , oops, sorry about that!  We had a couple of messy merges to update work Jeremy had done a couple years ago to E3SM master, and this fell through the cracks of that process.  I've pushed a commit restoring that line.  We can squash it into the commit that deleted it if that ends up being preferable.  Thanks for helping review this.

billsacks commented:
@matthewhoffman I think you need to do more than just put that call in: much of the code now inlined in `cime_run` duplicates the code in `cime_run_glc_setup_send`. This needs to be reconciled by removing the inlined code in `cime_run` and modifying the code in `cime_run_glc_setup_send` accordingly.

billsacks commented on cime/src/drivers/mct/main/prep_glc_mod.F90:
```diff
@@ -498,6 +641,33 @@ subroutine prep_glc_merge( l2x_g, fractions_g, x2g_g )
+    do eoi = 1,num_inst_ocn
+      o2x_ox => component_get_c2x_cx(ocn(eoi))
+      call seq_map_map(mapper_So2g, o2x_ox, o2x_gx(eoi), &
+                       fldlist='So_blt:So_bls:So_htv:So_stv:So_rhoeff',norm=.true.)
```
Actually, this may not be as important or feasible given that these specific fields are used in the new routines below. A compromise would be to introduce a module-level parameter in prep_glc_mod giving this field list, rather than having it duplicated between here and a second place below.

matthewhoffman replied:
I did your first suggestion, because it seemed to make sense to me after looking over that implementation in seq_flds_mod.  Changed in 84e1c872b35a13d9540b507c1d381d08202ba35c.

billsacks commented on cime/src/drivers/mct/main/prep_glc_mod.F90:
```diff
@@ -607,6 +777,156 @@ end subroutine prep_glc_map_one_state_field_lnd2glc
+    index_x2g_So_bls =    mct_avect_indexra(x2g_gx,'So_bls',perrwith='quiet')
+    index_x2g_So_htv =    mct_avect_indexra(x2g_gx,'So_htv',perrwith='quiet')
+    index_x2g_So_stv =    mct_avect_indexra(x2g_gx,'So_stv',perrwith='quiet')
+    !index_x2g_So_rhoeff = mct_avect_indexra(x2g_gx,'So_rhoeff',perrwith='quiet')
```
rhoeff is remapped above but then commented out here. Please clean this up.

matthewhoffman replied:
Fixed in 733a46925b5580e5888bdfa417add9916e7cf2df

billsacks commented on cime/src/drivers/mct/main/prep_glc_mod.F90:
```diff
@@ -607,6 +777,156 @@ end subroutine prep_glc_map_one_state_field_lnd2glc
+      outIceHeatFlux(n) =                 0.0_r8
+    end do
+
+    !Fyke: hardwire in some *reasonable* values here to test things out.
```
More commented-out code to remove.

matthewhoffman replied:
Fixed in 733a46925b5580e5888bdfa417add9916e7cf2df

billsacks commented on cime/src/drivers/mct/main/prep_glc_mod.F90:
```diff
@@ -607,6 +777,156 @@ end subroutine prep_glc_map_one_state_field_lnd2glc
+      !     end do
+
+    !...calculate fluxes...
+    call compute_melt_fluxes(oceanTemperature,&
```
Please use keyword arguments when you have an argument list this long, to avoid subtle bugs with out-of-order arguments.

matthewhoffman replied:
Fixed in 733a46925b5580e5888bdfa417add9916e7cf2df

matthewhoffman commented:
@billsacks , ahh, so I'm guessing `cime_run_glc_setup_send` was a routine added after Jeremy started his branch that moved out a bunch of code into a new subroutine.  I will work on sorting through those differences tomorrow and let you know when I have it worked out.

billsacks commented on cime/src/drivers/mct/main/prep_glc_mod.F90:
```diff
@@ -607,6 +777,156 @@ end subroutine prep_glc_map_one_state_field_lnd2glc
+                             outIceHeatFlux,&
+                             gsize)
+
+    !do n=1,gsize
```
More commented-out code to remove.

matthewhoffman replied:
Fixed in 733a46925b5580e5888bdfa417add9916e7cf2df

billsacks commented on cime/src/drivers/mct/main/prep_glc_mod.F90:
```diff
@@ -1108,4 +1443,176 @@ function prep_glc_get_mapper_Fl2g()
+!
+!-----------------------------------------------------------------------
+
+  subroutine compute_melt_fluxes( &
```
This routine does not match the style used throughout cime's Fortran code, and should be cleaned up. (Granted, the existing code style isn't fully consistent, but this routine differs more substantially than normal.):

- The above documentation should be edited to match more typical documentation styles in cime

- Whitespace should be cleaned up to match typical indentation (e.g., indentation of continuation lines for the argument list)

- Variable declaration lines should be edited to match the common style: `real(r8)` rather than `real (kind=r8)`, separate declaration lines for each variable, `real(r8) :: foo(:)` rather than using the `dimension` keyword, etc.

matthewhoffman replied:
@billsacks , I find this routine to be easier to read than most of CIME, so I'd prefer to keep it as stylistically written, even if it differs from other routines, particularly given it's a specialized routine that is not at all central to the workings of the module.  But if you feel particularly strongly about this, let me know.

billsacks commented on cime/src/drivers/mct/main/prep_glc_mod.F90:
```diff
@@ -1108,4 +1443,176 @@ function prep_glc_get_mapper_Fl2g()
+
+  !--------------------------------------------------------------------
+
+  end subroutine compute_melt_fluxes !}}}
```
Remove comment on end of this line.

matthewhoffman replied:
Fixed in 733a46925b5580e5888bdfa417add9916e7cf2df

billsacks commented on cime/src/drivers/mct/main/prep_glc_mod.F90:
```diff
@@ -1108,4 +1443,176 @@ function prep_glc_get_mapper_Fl2g()
+      outIceHeatFlux(iCell) = outIceHeatFlux(iCell) &
+        - iceHeatFluxCoeff*(iceTemperature(iCell) - outInterfaceTemperature(iCell))
+
+      if (iCell==72) then
```
Please remove this block of diagnostic output.

matthewhoffman replied:
Fixed in 733a46925b5580e5888bdfa417add9916e7cf2df

billsacks commented on cime/src/drivers/mct/main/prep_ocn_mod.F90:
```diff
@@ -1066,6 +1110,8 @@ subroutine prep_ocn_merge( flux_epbalfact, a2x_o, i2x_o, r2x_o, w2x_o, g2x_o, xa
        end if
     end do
+    !Experimental: call subshelf boundary flux code here.  Note this is called on atmospheric timestep.
+    call prep_glc_calculate_subshelf_boundary_fluxes
```
This is called from both here and cime_run. Is that duplication intentional? And if this is here intentionally: is there any precedent for calling one component's prep routine from another component's prep routine? I'm concerned that doing this makes it much harder to follow the driver sequencing logic.

matthewhoffman replied:
Removed in 284ab38dc40dd4dae45f093c0d8e391e24864d74.

billsacks commented on cime/src/drivers/mct/main/seq_rest_mod.F90:
```diff
@@ -221,6 +226,15 @@ subroutine seq_rest_read(rest_file, infodata, &
+
+       if (glc_present) then
+          gsmap         => component_get_gsmap_cx(glc(1))
+          x2gacc_gx     => prep_glc_get_x2gacc_gx()
```
This variable is only allocated in `prep_glc_mod` if `ocn_c2_glc` is true. Should this block of code be inside a `ocn_c2_glc` conditional?

matthewhoffman replied:
@billsacks , yes, I think you are correct here and below.  However, ocn_c2_glc is a module variable in cime_comp_mod, so it would have to be added to the calling signature for both `seq_rest_read` and `seq_rest_write`.  This would break with the convention of those routine calling signatures, but I don't see a good alternative.  Locally duplicating the logic defining ocn_c2_glc from the info in `infodata` would not be a good idea.  I  am fine with adding ocn_c2_glc to the calls to `seq_rest_read` and `seq_rest_write`, but I wanted to see if that seemed acceptable.

cc: @rljacob

billsacks commented on cime/src/drivers/mct/main/seq_rest_mod.F90:
```diff
@@ -480,6 +494,15 @@ subroutine seq_rest_write(EClock_d, seq_SyncClock, infodata, &
           end if
+          if (glc_present) then
+             gsmap         => component_get_gsmap_cx(glc(1))
+             x2gacc_gx => prep_glc_get_x2gacc_gx()
```
Same comment as above: This variable is only allocated in `prep_glc_mod` if `ocn_c2_glc` is true. Should this block of code be inside a `ocn_c2_glc` conditional?

billsacks commented on cime/src/drivers/mct/shr/seq_flds_mod.F90:
```diff
@@ -2292,6 +2292,7 @@ subroutine seq_flds_set(nmlfile, ID, infodata)
     call seq_flds_add(g2x_states_to_lnd,trim(name))
     call seq_flds_add(x2l_states,trim(name))
     call seq_flds_add(x2l_states_from_glc,trim(name))
+    call seq_flds_add(x2o_states,trim(name))
```
This should be enclosed in a conditional, `if (trim(cime_model) == 'e3sm') then`

matthewhoffman replied:
Fixed in 59a26f5ec4f1598113a3e68d7f7498c8fa108005

billsacks commented on cime/src/drivers/mct/shr/seq_flds_mod.F90:
```diff
@@ -2394,6 +2395,148 @@ subroutine seq_flds_set(nmlfile, ID, infodata)
     call set_glc_elevclass_field(name, attname, longname, stdname, units, x2l_fluxes_from_glc, &
          additional_list = .true.)
 
+     name = 'So_blt'
```
This block of added fields should be enclosed in a conditional, `if (trim(cime_model) == 'e3sm') then`

matthewhoffman replied:
Fixed in 59a26f5ec4f1598113a3e68d7f7498c8fa108005

billsacks commented:
> @billsacks , ahh, so I'm guessing `cime_run_glc_setup_send` was a routine added after Jeremy started his branch that moved out a bunch of code into a new subroutine. I will work on sorting through those differences tomorrow and let you know when I have it worked out.

Yes, that was introduced fairly recently.

billsacks commented:
@jgfouca maybe this was already your plan, but I'd like to request that, once this comes to e3sm master, it should be split into ESMCI's cime as its own PR, without anything else. That will facilitate testing on our end, because we'll want to test this fairly extensively in CESM before bringing it to cime master.

jgfouca commented:
@billsacks , unfortunately, that's not possible. But I can arrange it so there's very little else (by doing the next split very soon after this goes in).

billsacks commented:
> @billsacks , unfortunately, that's not possible. But I can arrange it so there's very little else (by doing the next split very soon after this goes in).

Okay, thanks.

rljacob commented:
To minimize the extra stuff, could you do a split right before this is merged?  And then a second split right after should have only these changes?

jgfouca commented:
@rljacob , in order for my scripts to be able to correctly auto-resolve all the false conflicts flagged by git, splits and merges must alternate.

matthewhoffman commented:
@jgfouca , what about if we moved the CIME-related commits into a PR that goes straight into CIME, and then pull it into E3SM with one of your merges?  Then we'd add the E3SM-specific on in a separate E3SM PR.

rljacob commented:
@matthewhoffman yes a PR to CIME first would be best.

jgfouca commented:
@matthewhoffman , that would be fine, but it would take some skill with Git to ensure that it was done correctly. I plan on doing the next "split" (E3SM -> CIME) on Monday, so not long from now.

billsacks commented:
From discussing this with @rljacob and @jgfouca - I'm fine with the initial plan of keeping these changes in e3sm and then having @jgfouca do an e3sm -> cime split shortly afterwards.

rljacob commented:
So don't make a separate PR to ESMCI/CIME.  We'll continue with reviewing/integrating this PR.

Also if the git log for a cime file is confusing on this end, try looking at it in a clone of ESMCI/cime.

matthewhoffman commented on cime/src/drivers/mct/main/cime_comp_mod.F90:
```diff
@@ -2523,8 +2546,94 @@ subroutine cime_run()
+                                           !unsafe to do it here.
+          endif
+
+          call prep_glc_accum(lnd_c2_glc, timer='CPL:glcprep_accum') !accum x2g_g fields here into x2g_gacc, along with l2gacc_lx
```
@JeremyFyke , I am assuming that this call to `prep_glc_accum` needs to be here so that accumulation will occur for coupling fields from both LND and now OCN.  However, there already is a call to  `prep_glc_accum` from routine `cime_run_lnd_recv_post` in this file at ~line 3831.  I am guessing that calling `prep_glc_accum` from multiple places is a bad idea as it could cause over-accumulation.  My intution would be to delete the call to `prep_glc_accum` in `cime_run_lnd_recv_post` and keep it here.  But I'm not sure if that generates any other problems.  Do you happen to have recollection of this or thoughts on it?

cc: @billsacks

billsacks replied:
If you browse through the new version of `cime_run` (at least on cime master), you'll see that very little is now done directly in `cime_run`: that now mainly farms out to other subroutines. I think the same should be true for glc setup-send. (I'm not sure if that answers your question, though.)

matthewhoffman commented on cime/src/drivers/mct/main/cime_comp_mod.F90:
```diff
@@ -2523,8 +2546,94 @@ subroutine cime_run()
+          if (lnd_c2_glc) then
+             ! Note that l2x_gx is obtained from mapping the module variable l2gacc_lx
+             call prep_glc_calc_l2x_gx(fractions_lx, timer='CPL:glcprep_lnd2glc')
+                !...and then here is where Bill S. does the new in-coupler downscaling.  I changed this so it's downscaling
```
@billsacks , do you have any concerns with the change described in this comment?

billsacks replied:
Yes, I do have concerns about this, but I haven't fully digested these changes. This was what motivated me to ask you to reconcile this with the code in `cime_run_glc_setup_send`: it's hard for me to see what changes were intentional and what were unintentional. What I'd suggest is that you and @JeremyFyke reconcile these routines in the way that you'd like them, then I'll review that reconciliation and comment on anything that I think will cause trouble. Does that sound okay to you?

matthewhoffman replied:
@billsacks , specifically, what is going on here is that before you had the following order of operations:
1. `prep_glc_accum` (via `cime_run_lnd_recv_post`)  (accumulate fields on LND grid)
2. `prep_glc_accum_avg` (average fields on LND grid)
3. `prep_glc_calc_l2x_gx` (remap to GLC grid)
4. `prep_glc_mrg` (finalize remapped LND fields for usage by GLC)

@JeremyFyke 's changes reorder operations like this:
1. `prep_glc_calc_l2x_gx` (remap to GLC grid)
2. `prep_glc_mrg` (finalize remapped LND fields for usage by GLC)
3. `prep_glc_accum`  (accumulate fields)
4. `prep_glc_accum_avg` (average fields)

Looking at it closer, it looks like steps 3 and 4 in the new ordering are still happening on the LND grid, which would be a problem.

I believe the goal of the reordering was so there would be common `accum` and `accum_avg` steps for fields coming from both LND and OCN.  I think to do this, we'd have to modify those routines to work on the GLC grid for the fields coming from LND instead of on the LND grid.  Alternatively, we could keep the old order of operations for the LND coupling fields and then  have different versions of the  `accum` and `accum_avg` routines for the fluxes coming from LND vs. OCN. 

Let me know if you have any thoughts on this.

matthewhoffman replied:
@billsacks , my questions were motivated by my attempts to reconcile them.  It turns out most of `cime_run_glc_setup_send` gets gutted to reconcile the changes due to the order of operations change described in my previous comment.  (It's not possible to just reorganize inside `cime_run_glc_setup_send` because some of the steps have to move outside the if statement enclosing the call to `cime_run_glc_setup_send`.)   If my previous comment about the order of operations is not specific enough, I don't mind attempting to reconcile `cime_run_glc_setup_send`, but it might not look much simpler than it already does.

billsacks replied:
I don't completely understand all of the new code / changes, but: At least in CESM, we want to accumulate the lnd2glc fields on the land grid, and then only do the regridding once per year - i.e., based on `glcrun_avg_alarm`. This is important for both performance and scientific reasons (the latter because the lnd -> glc regridding is not a linear operation, because of the conservation corrections that are done in this remapping). So, I'm not sure how the new code needs to be organized to achieve that, but that's an important requirement to maintain.

It seems like you need separate prep_glc_accum routines to (1) accumulate fields from lnd2glc (done after lnd_run), and (2) accumulate fields from ocn2glc (done after ocn_run). So you could have something like `prep_glc_accum_from_lnd` and `prep_glc_accum_from_ocn`. It's probably still fine to have a single `prep_glc_accum_avg` routine, because that should only be needed right before running glc.

Does that help at all?

Regardless of what's done here, it would still be good to maintain the new cime_run design of having most work done by subroutines rather than having big blocks of code inlined in there - even if this requires adding one or two new subroutines holding the new code that has been added in cime_run.

matthewhoffman replied:
@billsacks , thanks for the input on this - that is actually very helpful.  Knowing that maintaining the existing functionality of accumulating lnd2glc fields on the land grid is a requirement makes it clear that the accumulation changes currently in the PR will not work.  Thanks for pointing out the scientific reason why that's not ok - that makes sense to me.  

I've been coming to the same conclusion that separate prep_glc_accum routines will be needed to handle this situation.  I was looking around for an analogous situation to what we have here elsewhere in the model, and I don't think there is one.  Namely, a component receiving fluxes from two other components that have different coupling intervals from one another.  As far as I can tell no other component has this situation where accumulation has to happen differently for different components from which it receives information.

I'm beginning to have a sense of how this area of the code needs to be refactored.  I'll try to draft up and test changes, and then I'll ask you to look over the proposal.  Might take a few days.

billsacks replied:
Thanks a lot, @matthewhoffman - this sounds good.

I'm thinking that the ocean receives and accumulates fields from components with different coupling frequencies (atm vs. rof), and in that case it looks like it accumulates fields on the ocn grid. But yeah, with the custom remapping we do from lnd -> glc, I don't think that'll work, and we need something different.

For what it's worth, I ran this by Mariana:

> It seems like you need separate prep_glc_accum routines to (1) accumulate fields from lnd2glc (done after lnd_run), and (2) accumulate fields from ocn2glc (done after ocn_run). So you could have something like prep_glc_accum_from_lnd and prep_glc_accum_from_ocn. 

and she agreed that that sounds like the right way to go.

matthewhoffman replied:
@billsacks , I've restructured the code in cime_comp_mod.F90 to deal with issues we've been discussing.  I've tested to make sure it compiles and the subshelf melt flux calculation still works, but I have not yet done more extensive testing.  I will work on your other code cleanup requests next and then do more thorough testing.  But I wanted to point out the two new commits I added now, in case you want to quickly look them over and make sure they address your major concerns.

billsacks commented on cime/src/drivers/mct/main/cime_comp_mod.F90:
```diff
@@ -3651,6 +3676,47 @@ subroutine cime_run_atmocn_setup(hashint)
+                                           !unsafe to do it here.
+          endif
+
+          call prep_glc_accum_ocn(timer='CPL:glcprep_accum_ocn') !accum x2g_g fields here into x2g_gacc, along with l2gacc_lx
```
Should this call be in a conditional beyond just `glc_present`? Maybe `ocn_c2_glc`?

matthewhoffman replied:
@billsacks , yes, good point.  I've added that in a new commit (312aa4fb348f8b5584ae612a4692a81860d5eb65).  I actually changed the conditional logic for the entire routine on further review, and I'm hoping to get blessing from @JeremyFyke on that change.

billsacks commented on cime/src/drivers/mct/main/cime_comp_mod.F90:
```diff
@@ -3741,24 +3807,27 @@ subroutine cime_run_glc_setup_send(lnd2glc_averaged_now)
+          endif ! lnd_c2_glc or ocn_c2_glc
+
+       else
+          call prep_glc_zero_fields()
```
Previously this call to `prep_glc_zero_fields` was only done if `lnd_c2_glc` was true. Now it's being done regardless of the value of `lnd_c2_glc` or `ocn_c2_glc`. I'm wondering if it should only be done if `lnd_c2_glc .or. ocn_c2_glc`. It may be okay as is, but can you confirm?

matthewhoffman replied:
@billsacks , funny, I had just been thinking last night that this was inadvertently changed and I need to take a closer look at it.  I'll admit I don't fully understand the requirements here, but the safest thing would be to retain the previous functionality as written, so I've added a commit that does that (312aa4fb348f8b5584ae612a4692a81860d5eb65).

matthewhoffman commented:
@billsacks , it has dawned on me that these new couplings will be active in any compsets that include both GLC and OCN, so I believe these change will affect your existing compsets that include CISM where OCN is also active.  If the grids don't overlap there won't be fluxes being communicated, but I believe the coupler would be looking for associated mapping files, etc., which is undesirable.

Do you have ideas/preferences about how to handle that?  The two options I can think of are 1) protecting all these things with "if e3sm" clauses or 2) adding some sort of namelist option or other logic to make the new OCN-GLC coupling optional.

billsacks commented:
> @billsacks , it has dawned on me that these new couplings will be active in any compsets that include both GLC and OCN, so I believe these change will affect your existing compsets that include CISM where OCN is also active. If the grids don't overlap there won't be fluxes being communicated, but I believe the coupler would be looking for associated mapping files, etc., which is undesirable.
> 
> Do you have ideas/preferences about how to handle that? The two options I can think of are 1) protecting all these things with "if e3sm" clauses or 2) adding some sort of namelist option or other logic to make the new OCN-GLC coupling optional.

Yes, I think this relates to my earlier comment, right?:

> One thing I'm not sure about is: Should the settings of ocn_c2_glc, glcshelf_c2_ocn, and glcshelf_c2_ice in cime_comp_mod be enclosed in an if (trim(cime_model) == 'e3sm') conditional, setting them to false for cesm for now?

I don't have strong feelings about protecting these with 'if e3sm' or a namelist flag. I'm fine with just using an 'if e3sm' conditional if you don't have any need for turning these couplings off yourself, because I assume that will be simpler for you.

My initial thinking is that, if those three logicals are all false, then that should be sufficient to prevent any of the new code from running - but I'm not sure if that's 100% true.

(By the way: I saw your other comments, and will try to look at them more closely soon.)

matthewhoffman commented:
@billsacks , thanks for the feedback.  Yes, using an 'if e3sm' conditional is easier.  In the long run we might want something more flexible, but I'd rather get what we have working now and figure out adjustments down the road if they end up being needed.  I would say you can hold off on further code review until I've had time to address the remaining issues and do further testing on our end.  It will probably be early-mid next week before that point.

matthewhoffman commented on cime/src/drivers/mct/main/prep_glc_mod.F90:
```diff
@@ -290,31 +398,49 @@ end subroutine prep_glc_set_g2x_lx_fields
+       l2gacc_lx_cnt = l2gacc_lx_cnt + 1
+    end if
+
+    if (glc_present) then
```
fixed in fdeb032e0c5a29981caf2284356250184ee22e6d

matthewhoffman commented on cime/src/drivers/mct/main/prep_glc_mod.F90:
```diff
@@ -342,6 +470,21 @@ subroutine prep_glc_accum_avg(timer)
+    do egi = 1,num_inst_glc
+       ! temporary formation of average
+       if (x2gacc_gx_cnt > 1) then
+	  call mct_avect_avg(x2gacc_gx(egi), x2gacc_gx_cnt)
```
I believe this is fixed in fdeb032e0c5a29981caf2284356250184ee22e6d

matthewhoffman commented:
@jhkennedy , if you are able to do, try testing this branch with your SMB runs.  Feel free to contact me by email to coordinate the best way to do that based on what version of the code you are currently working with.

jhkennedy commented:
@matthewhoffman I rebased my changes to get the AISGIS grid working onto this branch and tried to run out 20yrs from a cold start, but I am getting run-time errors like: 

```
 576: Rank 576 [Mon Apr 29 09:13:21 2019] [c0-3c1s6n2] application called MPI_Abort(MPI_COMM_WORLD, 1001) - process 576
  78:  (prep_glc_calc_l2x_gx) ERROR: Flux fields other than Flgl_qice
  78:   currently are not handled in lnd2glc remapping.
  78:  (Attempt to handle flux field <Fogx_qiceli>.)
  78:  Substantial thought is needed to determine how to remap other fluxes
  78:  in a smooth, conservative manner.
```

Everything built fine though, for: `-compset IGCLM45_MLI -res ne30_oECv3_aisgis`

matthewhoffman commented:
@billsacks , we (@jonbob and myself) have finished reworking a few things on our end, and this PR is ready for another round of review from you.  To allow for compsets in E3SM where both OCN and GLC are active but this new OCN-GLC coupling is not desired (e.g. Greenland), we added logic to only enable this new coupling if the OCN model configuration requests it.  That effectively eliminates the need to identify if the CIME model is E3SM or CESM.

I believe I've addressed all of your helpful concerns at this point, but let me know if you have any remaining (or new) issues.

@jhkennedy , if you can, please try testing this updated branch with your SMB runs.

matthewhoffman commented on cime/src/drivers/mct/main/cime_comp_mod.F90:
```diff
@@ -366,6 +366,8 @@ module cime_comp_mod
   ! control flags
   !----------------------------------------------------------------------------
 
+  character(len=cs) :: cime_model    ! model invoking CIME (cesm or e3sm)
```
Good catch - I've removed it in a new commit.

matthewhoffman commented:
@billsacks , thanks for looking things over again, and I'm glad we are converging towards success!  I've pushed 2 new commits that address your 2 remaining concerns.  Let us know if you see anything else that concerns you.

@rljacob , @billsacks , @jgfouca , I believe the plan discussed some weeks ago was to move this PR directly to the CIME repo.  Is that still the preferred path forward once Bill officially approves the changes?

billsacks commented:
> @rljacob , @billsacks , @jgfouca , I believe the plan discussed some weeks ago was to move this PR directly to the CIME repo. Is that still the preferred path forward once Bill officially approves the changes?

I think we said that @jgfouca would do a split to ESMCI/cime very soon after this comes in, and would try to time things such that that split includes little (if anything) besides this change. Does that still sound right?

jhkennedy commented:
@matthewhoffman I rebased again for the AISGIS grid and tried to run out 20yrs from a cold start, but I am still getting run-time errors like:
```
2:  (prep_glc_calc_l2x_gx) ERROR: Flux fields other than Flgl_qice 
2:   currently are not handled in lnd2glc remapping. 
2:  (Attempt to handle flux field <Fogx_qiceli>.) 
2:  Substantial thought is needed to determine how to remap other fluxes 
2:  in a smooth, conservative manner.
```

matthewhoffman commented:
@jhkennedy , thanks for testing that again.  I looked into it, and I'd like @billsacks' opinion on how to deal with this.  The problem is in subroutine prep_glc_calc_l2x_gx() in cime/src/drivers/mct/main/prep_glc_mod.F90.  There is a loop at what currently is lines 719-747 that loops over all x2g flux fields.  For any flux field besides 'Flgl_qice', this error is triggered.  That was fine before, because 'Flgl_qice' was the only x2g flux field, but with the new OCN-GLC coupling, we've added two new flux fields that are unrelated to LND.  (Note there is a second bit of logic in prep_glc_merge_lnd_forcing doing the same thing that will also need to be modified.)

I think the solution is to either:
1) modify this loop to ignore x2g flux fields coming from any component besides LND.  I'm not sure if this is possible or how to do that.
2) disable the error message and just silently ignore flux fields that are not 'Flgl_qice'.

@billsacks , do you have a preferred (or alternate) solution?

cc: @jonbob

billsacks commented:
I feel like the ideal solution would be (1). I think this could be done by introducing an auxiliary variable in seq_flds_mod that just holds the x2g fluxes from lnd (like `seq_flds_x2g_fluxes_from_lnd`), similar to what I think you did somewhere else in your changes.

However, I'm also truly fine with solution (2), since it will be significantly easier.

matthewhoffman commented:
@billsacks , I also think that solution 1 is better, and when I looked closer at the code I became more convinced, because those affected routines are doing other things to prep coupling fields for LND->GLC calculations and it may cause problems if those operations occur on fields not coming from LND (or at least is wasted computation).  I found that there also is a list of state fields being looped over, and this also should be restricted to state fields coming from LND.  Making the new lists of state and flux fields just coming from LND is straightforward, as you pointed out.

However, I ran into one problem here:
https://github.com/E3SM-Project/E3SM/blob/f3ed491e04e72c8102081f8840c4de42587894a3/cime/src/drivers/mct/main/prep_glc_mod.F90#L590
I am assuming there is no way to restrict this calculation:
```
nflds = mct_aVect_nRattr(x2g_g)
```
to only be for fields coming from LND, so this comparison:
```
if (nflds /= (num_flux_fields + num_state_fields)) then
```
 will cause an error if the flux and/or state field lists are restricted to only fields coming from LND.

My proposal would be to eliminate that comparison and instead allocate `mrgstr` with size `(num_flux_fields + num_state_fields)`.  `mrgstr` is only allocated here to write out a list of LND->GLC coupling fields to the log file, so these seems like a reasonable solution to me.

Let me know what you think.

billsacks commented:
@matthewhoffman sounds great - thanks for giving this some careful thought. Your proposal sounds fine to me.

matthewhoffman commented:
@jhkennedy , I believe I have pushed a fix for the errors you have been seeing.  Can you try testing this branch again?  Thanks for your continued help in making sure these new changes won't break the SMB work you've been doing!

jhkennedy commented:
@matthewhoffman great work! I'll kick off a run to test it this afternoon.

jhkennedy commented:
@matthewhoffman looks like those last changes did it! I was able to run a 20 year cold start and get the expected SMB fields out with the new AISGIS grid. So I'd say everything looks good to me.

stephenprice commented:
Thanks for the udpate and testing @jhkennedy. Maybe we can tack on some brief discussion of where to go next with this during our discussion on Monday?

jhkennedy commented:
@stephenprice I think that's a good idea.

rljacob commented:
Note, @jgfouca sent our latest CIME updates (on master) to the upstream on May 14.

jonbob commented:
@rljacob - does that change when we should merge this? We had been planning to merge to next today

rljacob commented:
No it doesn't.  Today is a good time to merge.  The next split will probably have just this PR in it.

jonbob commented:
merged to next

rljacob commented:
Testing against baselines shows diffs in compsets that have an active ice sheet model.  I assume that is expected.   But there are also diffs in the X cases.   Not sure about those.

jonbob commented:
@rljacob - yes, I'm looking into that

jonbob commented:
@rljacob - we did add fields into the cpl, and that changes the values that the dead models pass around. So I believe the diffs in the X cases are OK

rljacob commented:
You're right.  The values are related to the number of fields.

jonbob commented:
merged to master
